### ISSUE TITLE
Building Input Functions with tf.estimator

### ISSUE THREAD

**Issue by killelder** [killelder/tensorflownote#7]
https://www.tensorflow.org/get_started/input_fn

-------------------------------------------------------
這篇主要講如何建立input_fn for estimator

----------------------------------------------------------
Custom Input Pipelines with input_fn
建立自己的input_fn

可以在input_fn裡面做feature engineering, pre-processing

------------------------------------------------------------
Anatomy of an input_fn
解剖 input_fn

下面這段code說明了input_fn的骨架, 主要回傳
第一個feature_cols: input data, 是一個dict, 必須要包含key/value, map feature column name給tensor
應該是指先指定好那些名字, 等一下那些key 對應到相對應的tensor

第二個tensor: output data, 直接是一個tensor
在那裏面, 可以做data pre-processing, 例如把不要的data拿掉, 或feature scaling

ex. 1-1
def my_input_fn():

    #Preprocess your data here...

    #...then return 1) a mapping of feature columns to Tensors with
    #the corresponding feature data, and 2) a Tensor containing labels
    return feature_cols, labels 

-----------------------------------------------------------------
Converting Feature Data to Tensors

如果使用pandas或numpy,
可以使用下面的方法轉換建構input_fn

ex. 1-2-1
import numpy as np
#numpy input_fn.
my_input_fn = tf.estimator.inputs.numpy_input_fn(
    x={"x": np.array(x_data)},
    y=np.array(y_data),
    ...)

ex. 1-2-2
import pandas as pd
#pandas input_fn.
my_input_fn = tf.estimator.inputs.pandas_input_fn(
    x=pd.DataFrame({"x": x_data}),
    y=pd.Series(y_data),
    ...)

對於Sparse data, (大部分為0的data), 會更希望使用SparseTensor,
需要以下三個arguments,
1. dense_shape : input矩陣的形狀, 用一個list表示
2. indices : 註記那些位置是nozeros, 用一個list表示
3. values: 在對應indices的位置, 塞入values
用下面的範例就可以看的懂了

ex. 1-3
sparse_tensor = tf.SparseTensor(indices=[[0,1], [2,4]],
                                values=[6, 0.5],
                                dense_shape=[3, 5])

>> [[0, 6, 0, 0, 0]
       [0, 0, 0, 0, 0]
       [0, 0, 0, 0, 0.5]]

要注意的是, 當estimator使用input_fn的時候, 是接收整個input_fn的function object,
不是接收return

ex. 1-4
classifier.train(input_fn=my_input_fn, steps=2000)
classifier.train(input_fn=my_input_fn(training_set), steps=2000) XXXXXX是錯的

如果真的想要代parameters, 要使用wrappers

ex. 1-5
def my_input_fn(data_set):
  ...

def my_input_fn_training_set():
  return my_input_fn(training_set)

classifier.train(input_fn=my_input_fn_training_set, steps=2000)

或是可以使用functools.partial

ex. 1-6
classifier.train(
    input_fn=functools.partial(my_input_fn, data_set=training_set),
    steps=2000)

第三個選項是使用lambda

ex. 1-7
classifier.train(input_fn=lambda: my_input_fn(training_set), steps=2000)

最後可以使用tf.estimator.inputs, 建立numpy和pandas的input_fn
可以把epoch, shuffle這些參數都設定好

ex. 1-8
import pandas as pd

def get_input_fn_from_pandas(data_set, num_epochs=None, shuffle=True):
  return tf.estimator.inputs.pandas_input_fn(
      x=pdDataFrame(...),
      y=pd.Series(...),
      num_epochs=num_epochs,
      shuffle=shuffle)
import numpy as np

def get_input_fn_from_numpy(data_set, num_epochs=None, shuffle=True):
  return tf.estimator.inputs.numpy_input_fn(
      x={...},
      y=np.array(...),
      num_epochs=num_epochs,
      shuffle=shuffle)

------------------------------------------------------------------
A Neural Network Model for Boston House Values

一樣先下載資料
ex. 2-1
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import itertools

import pandas as pd
import tensorflow as tf

tf.logging.set_verbosity(tf.logging.INFO)
COLUMNS = ["crim", "zn", "indus", "nox", "rm", "age",
           "dis", "tax", "ptratio", "medv"]
FEATURES = ["crim", "zn", "indus", "nox", "rm",
            "age", "dis", "tax", "ptratio"]
LABEL = "medv"

training_set = pd.read_csv("boston_train.csv", skipinitialspace=True,
                           skiprows=1, names=COLUMNS)
test_set = pd.read_csv("boston_test.csv", skipinitialspace=True,
                       skiprows=1, names=COLUMNS)
prediction_set = pd.read_csv("boston_predict.csv", skipinitialspace=True,
                             skiprows=1, names=COLUMNS)

-------------------------------------------------------------
Defining FeatureColumns and Creating the Regressor
建立regressor, 跟之前的範例都差不多
ex. 2-2
regressor = tf.estimator.DNNRegressor(feature_columns=feature_cols,
                                      hidden_units=[10, 10],
                                      model_dir="/tmp/boston_model")

------------------------------------------------------------
Building the input_fn

ex. 2-3
def get_input_fn(data_set, num_epochs=None, shuffle=True):
  return tf.estimator.inputs.pandas_input_fn(
      x=pd.DataFrame({k: data_set[k].values for k in FEATURES}),
      y = pd.Series(data_set[LABEL].values),
      num_epochs=num_epochs,
      shuffle=shuffle)

num_epochs 訓練時設定為None, 測試時設定為1
shuffle 訓練時設定為True, 測試時設定為False
